### PR TITLE
DSP/HLE: Implement Source processing

### PR DESCRIPTION
**Pull request by MerryMage**
This should close a lot of audio related issues.

This PR performs per-source audio processing and reporting of audio status back to the application.

### CODE REVIEW

File: src/audio_core/hle/source.cpp [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+    }
+
+    if (config.rate_multiplier_dirty) {
+        config.rate_multiplier_dirty.Assign(0);
+        state.rate_multiplier = config.rate_multiplier;
+        LOG_TRACE(Audio_DSP, "source_id=%zu rate=%f", source_id, state.rate_multiplier);
+
+        if (state.rate_multiplier <= 0) {
+            LOG_ERROR(Audio_DSP, "Was given an invalid rate multiplier: source_id=%zu rate=%f", source_id, state.rate_multiplier);
+            state.rate_multiplier = 1.0f;
```
Comment (MerryMage) [inline src/audio_core/hle/source.cpp]: Should probably note in a comment that the real firmware just freaks out at this point.

File: src/audio_core/hle/source.h [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+    using Format = SourceConfiguration::Configuration::Format;
+    using InterpolationMode = SourceConfiguration::Configuration::InterpolationMode;
+    using MonoOrStereo = SourceConfiguration::Configuration::MonoOrStereo;
+
+    /// Internal representation of a buffer for our buffer queue
+    struct Buffer {
+        PAddr physical_address;
+        u32 length;
+        u8 adpcm_ps;
+        u16 adpcm_yn[2];
```
Comment (lioncash) [inline src/audio_core/hle/source.h]: This can be a std::array

File: src/audio_core/hle/source.h [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+    /// Resets internal state.
+    void Reset();
+
+    /**
+     * This is called once every audio frame. This performs per-source processing every frame.
+     * @param config The new configuration we've got for this Source from the application.
+     * @param adpcm_coeffs ADPCM coefficients to use if config tells us to use them (may contain invalid values otherwise).
+     * @return The current status of this Source. This is given back to the emulated application via SharedMemory.
+     */
+    SourceStatus::Status Tick(SourceConfiguration::Configuration& config, const s16_le adpcm_coeffs[16]);
```
Comment (lioncash) [inline src/audio_core/hle/source.h]:
```cpp
const s16_le (&adpcm_coeffs)[16]
```

File: src/audio_core/hle/source.h [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+        // Filter state
+
+        SourceFilters filters;
+
+    } state;
+
+    // Internal functions
+
+    /// INTERNAL: Update our internal state based on the current config.
+    void ParseConfig(SourceConfiguration::Configuration& config, const s16_le adpcm_coeffs[16]);
```
Comment (lioncash) [inline src/audio_core/hle/source.h]:
```cpp
const s16_le &(adpcm_coeffs)[16]
```

File: src/audio_core/hle/source.h [hunk 5]
```diff
@@ -0,0 +1,145 @@
...
+     * @param dest The QuadFrame32 to mix into.
+     * @param intermediate_mix_id The id of the intermediate mix whose gains we are using.
+     */
+    void MixInto(QuadFrame32& dest, size_t intermediate_mix_id) const;
+
+private:
+    const size_t source_id;
+    StereoFrame16 current_frame;
+
+    using DspBuffer = SourceConfiguration::Configuration::Buffer;
```
Comment (MerryMage) [inline src/audio_core/hle/source.h]: This is not used anywhere and can be removed.

**Comment by MerryMage** [comment]
Addressed the above comments.

File: src/audio_core/hle/source.cpp [hunk 6]
```diff
@@ -0,0 +1,320 @@
...
+        const unsigned num_channels = buf.mono_or_stereo == MonoOrStereo::Stereo ? 2 : 1;
+        switch (buf.format) {
+        case Format::PCM8:
+            state.current_buffer = Codec::DecodePCM8(num_channels, memory, buf.length);
+            break;
+        case Format::PCM16:
+            state.current_buffer = Codec::DecodePCM16(num_channels, memory, buf.length);
+            break;
+        case Format::ADPCM:
+            ASSERT(num_channels == 1);
```
Comment (MerryMage) [inline src/audio_core/hle/source.cpp]: `DEBUG_ASSERT`, perhaps?

**Comment by MerryMage** [comment]
Rebased.

File: src/audio_core/hle/source.cpp [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+        UNIMPLEMENTED();
+        break;
+    }
+
+    state.current_sample_number = 0;
+    state.next_sample_number = 0;
+    state.current_buffer_id = buf.buffer_id;
+    state.buffer_update = buf.from_queue;
+
+    LOG_TRACE(Audio_DSP, "Source::DequeueBuffer(): source_id=%zu buffer_id=%hu from_queue=%s current_buffer.size()=%zu",
```
Comment (bunnei) [inline src/audio_core/hle/source.cpp]: The function name should be automatically logged

File: src/audio_core/hle/source.cpp [hunk 8]
```diff
@@ -0,0 +1,320 @@
...
+        frame_position += size_to_copy;
+        state.next_sample_number += static_cast<u32>(size_to_copy);
+    }
+
+    state.filters.ProcessFrame(current_frame);
+}
+
+
+bool Source::DequeueBuffer() {
+    ASSERT(state.current_buffer.empty());
```
Comment (bunnei) [inline src/audio_core/hle/source.cpp]: ASSERT_MSG?

File: src/audio_core/hle/source.cpp [hunk 9]
```diff
@@ -0,0 +1,320 @@
...
+    const Buffer buf = state.input_queue.top();
+    state.input_queue.pop();
+
+    if (buf.adpcm_dirty) {
+        state.adpcm_state.yn1 = buf.adpcm_yn[0];
+        state.adpcm_state.yn2 = buf.adpcm_yn[1];
+    }
+
+    if (buf.is_looping) {
+        LOG_WARNING(Audio_DSP, "Looped buffers are unimplemented at the moment");
```
Comment (bunnei) [inline src/audio_core/hle/source.cpp]: Should this be an ERROR?

File: src/audio_core/hle/source.cpp [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+
+    switch (state.interpolation_mode) {
+    case InterpolationMode::None:
+        state.current_buffer = AudioInterp::None(state.interp_state, state.current_buffer, state.rate_multiplier);
+        break;
+    case InterpolationMode::Linear:
+        state.current_buffer = AudioInterp::Linear(state.interp_state, state.current_buffer, state.rate_multiplier);
+        break;
+    case InterpolationMode::Polyphase:
+        // TODO(merry): Implement polyphase interpolation
```
Comment (bunnei) [inline src/audio_core/hle/source.cpp]: Should this be logged some way so we have a sense of where this is needed?

File: src/audio_core/hle/source.cpp [hunk 11]
```diff
@@ -0,0 +1,320 @@
...
+
+    size_t frame_position = 0;
+
+    state.current_sample_number = state.next_sample_number;
+    while (frame_position < current_frame.size()) {
+        if (state.current_buffer.empty() && !DequeueBuffer()) {
+            break;
+        }
+
+        size_t size_to_copy = std::min(state.current_buffer.size(), current_frame.size() - frame_position);
```
Comment (bunnei) [inline src/audio_core/hle/source.cpp]: const

**Comment by bunnei** [comment]
Left some minor comments, code looks fine to me after they are addressed

File: src/audio_core/hle/source.cpp (hunk 10 above)
Comment (MerryMage) [inline src/audio_core/hle/source.cpp]: It's the best quality mode to use and hence it's used everywhere; I haven't gotten an accurate enough emulation (or figured out a nice way to consistently extract the coeffcients from the firmware).

**Comment by raven02** [comment]
Hopefully this can help resolving the sound delay issue as well .

File: src/audio_core/hle/source.cpp (hunk 10 above)
Comment (bunnei) [inline src/audio_core/hle/source.cpp]: OK... yeah, perhaps we don't need/want the spam then

Comment (MerryMage) [inline src/audio_core/hle/source.cpp]: Mhm! Sorry for not directly answering your question;;

**Comment by MerryMage** [comment]
Addressed @bunnei's comment above.

@raven02: I don't know about that. Could you perhaps open an issue?

**Comment by MerryMage** [comment]
Squashed.

**Comment by bunnei** [comment]
LGTM

**Comment by raven02** [comment]
@MerryMage .I will open an issue to cover it and test out this commit when merged .

**Pull request merged**